### ISSUE TITLE
Xml reader suddenly closes

### ISSUE THREAD

**Issue by djpeluca** [monitorjbl/excel-streaming-reader#141]
Hi, I'm having a problem when building an Iterator on top of your iterator, I'm doing this because I have to build another type of Row for requirements. So basically i'm building an intermediate layer between the excel and the fanal tableRow required with this excel-streaming-reader. The iterator that I use i rowIterator(). I generated a hasnext() that only ask your hasnext() (with a Try catch that returns false) and a next() that uses the next() of the rowIterator() to grab the new row, make a for to iterate all the cells (I can't use a CellIterator because I don't want to skip blank cells) build my Custom Row and return that. It should work fine, as your rowIterator actually works fine, This is encapsulated in a class that implements an Iterator.

The problem is when I use that class outside, The stream closes at row 10000. At this point I have to say that this is for large files. 1 sheet with  1 millon rows and 12 columns. It works fine for not that large files.

at this time im testing with this setup (I also tested with different numbers and the row where I'm getting the error changes, so it's probably something about the memory that i'm not handling well . Is there something that I'm not seeing/understanding? Is there a time out or something? Can it be fixed? Thanks and regards

try(InputStream  file = new FileInputStream(new File(path));
            Workbook wb = StreamingReader.builder()
            .rowCacheSize(10000)
            .bufferSize(102400000)
            .sstCacheSize(148576000)
            .open(file))

ps: I'm getting this particulary problem when I'm trying to iterate and print all the cell with a printbuffer that limits itself to 1000 rows at a time.

Pa2: I'm actually don't know what is sstCacheSize for

**Comment by djpeluca** [comment]
Ok still testing on this and I get it's the RowCacheSize, so The stream closes on it's first read. If I use less rows in cache, it's falls when it's hits that, on the other hand, if I use the builder with more rowCacheSize (let's say 15000) I get to iterate those, but then it closes. I'm reading your code but I didn't get to the problem yet. Thanks.

**Comment by monitorjbl** [comment]
You've set a very high row cache size, buffer size and SST cache size. These options will add a ton of memory overhead, I'd recommend lowering them and seeing if that helps.

The row cache contrpols how far ahead in the XML stream the library reads. If you set this high, it will eagerly fetch rows up to the number you specified. The buffer size is used when extracting the workbook out to disk (which is required for the XML streamer to function).

The SST (Shared Strings Table) is a table of shared strings. An Excel workbook extracts repeated strings to a separate file to save space, so when the stream has to read out one of these strings it's got to fetch it from the SST. This library does not buffer the entire SST into memory, it instead serializes the SST out to a temp file and caches a maximum number of entries in memory at a time (with LRU evictions). The sstCacheSize controls how big that cache is.

By using giant buffer sizes you're effectively using this library like vanilla POI, which is a really bad idea. POI will always be better than this library in that regard; it buffers a ton of data into memory and is pretty fast to do any arbitrary action. This library is intended to reduce the memory overhead of *reading* Excel files only, and it does this by trading speed for lower memory usage.

**Comment by djpeluca** [comment]
Great, I get that and i'll keep it in mind. I'm actually getting other types of errors when using less cache size, I think I got to this known issue https://github.com/monitorjbl/excel-streaming-reader/issues/95 but also other issues, that I probably getting them because of that iterator on top of the iterator.

for instance, this setup

try(InputStream  file = new FileInputStream(new File(path));
            Workbook wb = StreamingReader.builder()
            .rowCacheSize(100)
            .bufferSize(10240)
            .sstCacheSize(1485)
            .open(file))

the "ontop" interator it let me print only 100 rows. So beside the cache size i'm loosing the reader when that first cache is consumed. I still don't get if i'm handling it wrong (like in wrong programming, or it's just a problem of reference and memory lost from java)

I know that your iterator works fine if I iterate when I create the iterator, what I don't know is who's fault is that the reader closes when I save the iterator as a variable in a class and build an iterator of my own on top of that. I mean, it is clear that the issue is how i'm handling it, but what I don't know if there is something that I can do to make it work or it just wont work conceptually. Thanks and regards

**Comment by djpeluca** [comment]
Ok! we got it! it was probably already obvius that it was an issue on how we where handle the "ontop" iterator. So now we got it right as we passed only your iterator on the ontop iterator class builder. We tried different stuff and I thought that we had already tried that, maybe we did but got other errors and got mixed up. Thanks, I will close the issue and try to not bother you anymore. Thanks for the extremely quick response. Congrats on the library
